[WIP] Modify mnemonic lib, fixes issue #161

LorenzoAlluminio opened the pull request:
Modification:

- lib/paratii.eth.wallet
substituted bitcore-mnemonic with [bip39](https://github.com/bitcoinjs/bip39)
substituted bitcore-lib with [ethereumjs-wallet/hdkey](https://github.com/ethereumjs/ethereumjs-wallet#hd-wallet-api)
- test/paratii.eth.wallet
added test to verify that the seed generation is [bip39 compliant] (https://github.com/bitcoin/bips/blob/master/bip-0039.mediawiki#test-vectors)

- Things to yet to implement
handle invalid mnemonics (https://github.com/bitcoinjs/bip39#reminder-for-developers)

I'm not sure about the quality and security of the code, please let me know what you think and if explaination is needed :)

Also I'm not sure about the dependencies changes in package.json, because we just use the module ```hdkey``` of the ethereumjs-wallet.

~~The failed test it's because I added the bip39-vector file for the test in the data folder, and the failing test says this:
```    // NOTE THIS WILL Trigger an error if the director test/data content changes.```
but I don't know how to fix it~~ moved the json file to another folder

jellegerbrandy commented on lib/paratii.eth.wallet.js:
```diff
@@ -15,21 +15,31 @@ export function patchWallet (wallet, config) {
-      var hdRoot = new Mnemonic(this._mnemonic).toHDPrivateKey().xprivkey
+
+      var seed = bip39.mnemonicToSeed(this._mnemonic, this._passphrase)
+      // saving for testing purpose
```
There are two reasons why this is not good:
* as a general rule, we should avoid having code only needed for testing in the library. 
* security: the seed (And the privatekeys) are the secrets here, and are to be encrypted with the passphrase. Saving them as properties on the wallet sort of defeats all of this.

jellegerbrandy replied:
(I realize we do the same with the `_mnemonic` property, which has the same risks. We probably should change that as well. Please let me know ifyou change this, because that will perhaps break things in the portal as well)

geckoslair commented on test/paratii.eth.wallet.js:
```diff
@@ -167,4 +167,22 @@ describe('paratii.eth.wallet: :', function () {
   })
+  it('eth.wallet.create() respect bip39 test vector', async function () {
+    var fs = require('fs')
+    var vector = JSON.parse(fs.readFileSync('test/testData/bip39-test-vector.json', 'utf8'))
```
for JSON data we can use also just the require instead of fs, maybe is more lighter!

jellegerbrandy commented on lib/paratii.eth.wallet.js:
```diff
@@ -50,17 +60,22 @@ export function patchWallet (wallet, config) {
     return this._mnemonic
   }
 
+  function setPassphrase (passphrase) {
```
Why do we need this?

jellegerbrandy commented on lib/paratii.eth.wallet.js:
```diff
@@ -15,21 +15,31 @@ export function patchWallet (wallet, config) {
+        var privkeyHex = generatedAddress.getPrivateKey().toString('hex')
 
-        var privkeyHex = privkeyBuf.toString('hex')
         if (privkeyBuf.length < 32) {
```
we do not need thesse checks anymore (they were a workaround around the buggy lib we used before). Can you remove them?

jellegerbrandy commented on package.json:
```diff
@@ -25,9 +25,9 @@
-    "bitcore-mnemonic": "^1.5.0",
+    "bip39": "^2.5.0",
     "es6-promise": "^4.2.4",
+    "ethereumjs-wallet": "^0.6.0",
```
I'm not in principel against using this, but we are trying to econmize on package size, and we can perhaps use https://www.npmjs.com/package/hdkey (I  mention that because ethereumjs-wallet uses that https://github.com/ethereumjs/ethereumjs-wallet/blob/master/package.json#L37 and it seems we are only using that part)

LorenzoAlluminio commented:
I've done a quick fix and done this changes:
- **REMOVED THIS._MNEMONIC** and all other object properties to ensure more security
- removed the setPassphrase method
- removed the useless ifs
- fixed the dependencies

I used the properties and the setPassphrase method to test that everything was going well. Now those tests are skipped.

Tomorrow I'll work on find a way of doing the tests and to add the passphrase ( note: now there is no passphrase)
